### PR TITLE
Tweaks to expressions to get complex index types to work

### PR DESCRIPTION
**Pull request by ckreibich**
The recent `CompositeHash` work in #1772 removed inconsistencies in the way we handled complex types for indexing, like the fact that we could index with a record that has a set member, but not with a set directly. This PR addresses the remaining bits of work around indexing and iteration to make such types work in practice. See the new test cases for examples. Essentially: if we can hash it, we can now index with it.

Caveats:
- If we explicitly do not want some of those index types, now would be a good time to ensure that
- I think this covers all functionality around indexing, but let me know if I missed anything
- As far as I can tell this doesn't affect ZAM, but here too shout if I'm missing something

### CODE REVIEW

File: src/Expr.cc [hunk 1]
```diff
@@ -4294,13 +4294,15 @@ InExpr::InExpr(ExprPtr arg_op1, ExprPtr arg_op2)
 
 	if ( op1->GetType()->Tag() == TYPE_PATTERN )
 		{
-		if ( op2->GetType()->Tag() != TYPE_STRING )
+		if ( op2->GetType()->Tag() == TYPE_STRING || op2->GetType()->Tag() == TYPE_TABLE )
 			{
```
Comment (0xxon) [inline src/Expr.cc:4298]: I might be missing something here - could you perhaps explain which case this covers?

Comment (0xxon) [reply src/Expr.cc:4298]: ah, ok, I think I got it. Let me just try one thing...

File: src/Expr.cc [hunk 2]
```diff
@@ -4294,13 +4294,15 @@ InExpr::InExpr(ExprPtr arg_op1, ExprPtr arg_op2)
 
 	if ( op1->GetType()->Tag() == TYPE_PATTERN )
 		{
-		if ( op2->GetType()->Tag() != TYPE_STRING )
+		if ( op2->GetType()->Tag() == TYPE_STRING || op2->GetType()->Tag() == TYPE_TABLE )
```
Comment (0xxon) [inline src/Expr.cc:4297]: Currently, this only checks if the second type is a table. This means that code like this:

```
local s1: set[count] = { 1, 5, 7 };
print ( /pat1/ in s1 ) ;
```

will happily run and print false.

I suspect we should catch this and do the same that happens when we, e.g., try to compare an integer with a table of patterns. In this case, the output of:

```
local s4: set[pattern] = { /pat1/ };
print (5 in s4);
```

is something along the lines of:

```
error in pattern and ./c.zeek, line 8: arithmetic mixed with non-arithmetic (pattern and 5)
error in ./c.zeek, line 8 and pattern: type mismatch (5 and pattern)
error in ./c.zeek, line 8: not an index type (5 in s4)
```

**Review by 0xxon — CHANGES_REQUESTED** [review]
This mostly looks great - thanks a lot. I love the expanded tests.

The only thing that I would change is that I think we should add a more stringent test for the type of the set when comparing a pattern with a set.

File: src/Expr.cc (hunk 2 above)
Comment (ckreibich) [reply src/Expr.cc:4297]: Nice catch — I'll get on it.

Comment (ckreibich) [reply src/Expr.cc]: I've reworked the logic in the `InExpr` constructor a bit further so the pattern case falls through to the general type-checking at the end of the function, which catches the above. I've expanded the corresponding test case, too.

**Review by 0xxon — APPROVED** [review]
thank you :)

**Pull request merged**